### ISSUE TITLE
A bug in sort! or delete!

### ISSUE THREAD

**Issue by ebelnikola** [andyferris/Dictionaries.jl#150]
Hi! 

I noticed quite non intuitive results of `sort!`  when it is applied to a dictionary after an entry was deleted. Consider the following piece of code:
```
a=Dictionary([1,2,3,4],[4,5,6,7])
delete!(a,4)
sort!(a)
```
I would expect that result of this is a dictionary of the following form:
```
3-element Dictionary{Int64, Int64}
 1 │ 4
 2 │ 5
 3 │ 6
```
However, the output is:
```
4-element Dictionary{Int64, Int64}
 1 │ 4
 2 │ 5
 3 │ 6
 4 │ #undef
```
I cannot access index 4. Trying to assign it some value returns an error, which is reasonable: 
```
IndexError("Dictionary does not contain index: 4")
``` 
But the println function can access it! The call `a .|> println` returns:
```
4
5
6
7
```
Function `length` sees that there are four values as well: `length(a)` returns 4. 

What is worse, deep copy of `a` restores index 4 and its value. You may see that `deepcopy(a)` returns:
```
4-element Dictionary{Int64, Float64}
 1 │ 4
 2 │ 5
 3 │ 6
 4 │ 7
```

**Comment by andyferris** [comment]
Thanks. This is definitely a big we need to fix.

**Comment by andyferris** [comment]
Fixed in 0.4.3

**Comment by andyferris** [comment]
Thanks @ebelnikola for reporting this!
